lib/client/alpn.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	rootCAs := x509.NewCertPool()
+
+	alpnUpgradeRequired := alpnproxy.IsALPNConnUpgradeRequired(req.WebProxyAddr, req.InsecureSkipVerify)
+
+	if alpnUpgradeRequired {
+		caCert, err := req.Client.GetClusterCACert(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if ok := rootCAs.AppendCertsFromPEM(caCert.GetTLSCA()); !ok {
+			return fmt.Errorf("failed to append cert from cluster's TLS CA Cert")
+		}
+	}
+
+	address, err := utils.ParseAddr(req.WebProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	key, err := GenerateRSAKey()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	currentUser, err := req.Client.GetCurrentUser(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	certs, err := req.Client.GenerateUserCerts(ctx, proto.UserCertsRequest{
+		PublicKey:              key.MarshalSSHPublicKey(),
+		Username:               currentUser.GetName(),
+		Expires:                time.Now().Add(time.Minute).UTC(),
+		ConnectionDiagnosticID: req.ConnectionDiagnosticID,
+		RouteToDatabase:        req.RouteToDatabase,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tlsCert, err := keys.X509KeyPair(certs.TLS, key.PrivateKeyPEM())
+	if err != nil {
+		return trace.BadParameter("failed to parse private key: %v", err)
+	}
+
+	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		InsecureSkipVerify:      req.InsecureSkipVerify,
+		RemoteProxyAddr:         req.WebProxyAddr,
+		Protocols:               protocols,
+		Listener:                req.Listener,
+		ParentContext:           ctx,
+		SNI:                     address.Host(),
+		Certs:                   []tls.Certificate{tlsCert},
+		RootCAs:                 rootCAs,
+		ALPNConnUpgradeRequired: alpnUpgradeRequired,
+		Middleware:              req.ProxyMiddleware,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		defer req.Listener.Close()
+		if err := lp.Start(ctx); err != nil {
+			log.WithError(err).Info("ALPN proxy stopped.")
+		}
+	}()
+
+	return nil
+}
+
+// ALPNCertChecker implements alpnproxy.LocalProxyMiddleware.
+// It has basic checks and supports adding custom checks based on the extracted Identity from the certificate.
+type ALPNCertChecker struct {
+	checkCerts func(lp *alpnproxy.LocalProxy) error
+}
+
+// NewALPNCertChecker creates a new ALPNCertChecker.
+func NewALPNCertChecker(certChecker func(*alpnproxy.LocalProxy) error) *ALPNCertChecker {
+	return &ALPNCertChecker{
+		checkCerts: certChecker,
+	}
+}
+
+// OnNewConnection is a callback triggered when a new downstream connection is
+// accepted by the local proxy.
+func (c *ALPNCertChecker) OnNewConnection(ctx context.Context, lp *alpnproxy.LocalProxy, conn net.Conn) error {
+	return trace.Wrap(c.checkCerts(lp))
+}
+
+// OnStart is a callback triggered when the local proxy starts.
+func (c *ALPNCertChecker) OnStart(ctx context.Context, lp *alpnproxy.LocalProxy) error {
+	return trace.Wrap(c.checkCerts(lp))
+}
```

Comment:
The Cert checker was introduced to monitor users cert during life time of the ALPN local proxy. Thus in case of TestConnection local proxy is spawned only for short period we can just pass a nil to the `LocalProxyConfig.Middleware` setting.